backend/security/src/main/kotlin/io/tolgee/security/ratelimit/RateLimitInterceptor.kt

```diff
@@ -0,0 +1,101 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.tolgee.security.ratelimit
+
+import io.tolgee.configuration.tolgee.RateLimitProperties
+import io.tolgee.dtos.cacheable.UserAccountDto
+import io.tolgee.security.authentication.AuthenticationFacade
+import org.springframework.core.Ordered
+import org.springframework.core.annotation.AnnotationUtils
+import org.springframework.stereotype.Component
+import org.springframework.web.method.HandlerMethod
+import org.springframework.web.servlet.HandlerInterceptor
+import org.springframework.web.servlet.HandlerMapping
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+@Component
+class RateLimitInterceptor(
+  private val authenticationFacade: AuthenticationFacade,
+  private val rateLimitProperties: RateLimitProperties,
+  private val rateLimitService: RateLimitService,
+) : HandlerInterceptor, Ordered {
+  override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
+    if (handler !is HandlerMethod) {
+      return super.preHandle(request, response, handler)
+    }
+
+    val policy = extractEndpointRateLimit(request, authenticationFacade.authenticatedUserOrNull, handler)
+    if (policy != null) {
+      rateLimitService.consumeBucket(policy)
+    }
+
+    return true
+  }
+
+  fun extractEndpointRateLimit(
+    request: HttpServletRequest,
+    account: UserAccountDto?,
+    handler: HandlerMethod
+  ): RateLimitPolicy? {
+    @Suppress("DEPRECATION") // TODO: remove for Tolgee 4 release
+    if (!rateLimitProperties.enabled) return null
+
+    val annotation = AnnotationUtils.getAnnotation(handler.method, RateLimited::class.java)
+      ?: return null
+
+    if (
+      (!annotation.isAuthentication && !rateLimitProperties.endpointLimits) ||
+      (annotation.isAuthentication && !rateLimitProperties.authenticationLimits)
+    ) {
+      return null
+    }
+
+    val matchedPath = request.getAttribute(HandlerMapping.BEST_MATCHING_PATTERN_ATTRIBUTE) as String
+    val pathVariablesMap = request.getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE) as Map<*, *>
+
+    val id = if (account?.id != null) "user.${account.id}" else "ip.${request.remoteAddr}"
+    val pathVariables = pathVariablesMap.values.iterator()
+
+    var bucketName = "endpoint.$id.${annotation.bucketName.ifEmpty { "${request.method} $matchedPath" }}"
+
+    // Include route parameters to discriminate major routes, but not minor routes.
```

Comment:
Do we have to manually handle the path? Isn't rate limiting single handler enough?

Reply:
I fetch the raw route definition without parameters in place, e.g. `/v2/projects/{id}/blah`, which will be common for all projects. The idea here is to make bucketing separate per project id, instead of being global.

Reply:
I understand, but it seems a bit too complicated to use to me. But if you think it's necessary, then keep it.

Reply:
The default behavior of the annotation will be the desired one 99% of the time. By default it'll discriminate against the 1st major route parameter, so the ratelimit is applied per-project, per-organization, etc. where applicable automatically.

The per-project limit can be very useful for orgs doing operations on many projects via a bot for example; it makes the ratelimit more permissive for these usecases without lowering the DoS protection provided by a rate-limiter.

The ability to configure it is left for more advanced case and for flexibility, such as for extremely expensive endpoints or to combat certain abuse patterns (e.g. comments, once community-based features are out).